rfcs/0000-trust-store-for-tls-and-ssh/README.md

```diff
@@ -0,0 +1,293 @@
+could be your own self-signed certificates, are considered untrustworthy by default.
+TLS communications against untrusted remote servers are aborted.
+
+#### Controller-level Trusted Certificates
+
+*NOTE:* this requires no changes on the controllers, as this is based on the ways
+that TLS surface the trust store. This RFC only formalizes it as a supported
+approach.
+
+To trust CAs that are not part of the root trusted CAs, the OS level trust store
+needs to be updated by mounting either a `ConfigMap` or `Secret` directly into the
+Flux Controllers.
+
+`Secret` example:
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: flux-trust-store
+  namespace: flux-system
+data:
+  customCA.pem: LS0tLS1CRUdJTiBDRVJUSUZJQ0FURS0tLS0tCk1JSUJoekNDQVMyZ0F3SUJBZ0lVZHNBdGlYM2dOMHVrN2RkeEFTV1lFL3RkdjB3d0NnWUlLb1pJemowRUF3SXcKR1RFWE1CVUdBMVVFQXhNT1pYaGhiWEJzWlM1amIyMGdRMEV3SGhjTk1qQXdOREUzTURneE9EQXdXaGNOTWpVdwpOREUyTURneE9EQXdXakFaTVJjd0ZRWURWUVFERXc1bGVHRnRjR3hsTG1OdmJTQkRRVEJaTUJNR0J5cUdTTTQ5CkFnRUdDQ3FHU000OUF3RUhBMElBQks3aC81RDhiVjkzTW1FZGh1MDJKc1M2dWdCOHM2UHpSbDNQVjR4czNTYnIKUk5ra001OSt4M2IwaVd4L2k3NnFQWXBOTG9pVlVWWFFtQTlZKzREYk14aWpVekJSTUE0R0ExVWREd0VCL3dRRQpBd0lCQmpBUEJnTlZIUk1CQWY4RUJUQURBUUgvTUIwR0ExVWREZ1FXQkJRR3lVaVUxUUVaaU1BcWpzbklZVHdaCjR5cDV3ekFQQmdOVkhSRUVDREFHaHdSL0FBQUJNQW9HQ0NxR1NNNDlCQU1DQTBnQU1FVUNJUUR6ZHR2S2RFOE8KMStXUlRaOU11U2lGWWNyRXo3Wm5lN1ZYb3VERUtxS0VpZ0lnTTRXbGJEZXVOQ0ticWhxait4WlYwcGEzcndlYgpPRDhFampDTVk2OVJNTzA9Ci0tLS0tRU5EIENFUlRJRklDQVRFLS0tLS0K
+```
+
+Patch required on the main `kustomization.yaml`:
+```yaml
+  - patch: |
+      - op: add
+        path: /spec/template/spec/containers/0/volumeMounts
+        value:
+          - name: tls-trust-store
+            mountPath: /etc/ssl/certs/ca-cert-flux.pem
+            subPath: customCA.pem
+            readOnly: true
+      - op: add
+        path: /spec/template/spec/volumes
+        value:
+          - name: tls-trust-store
+            secret:
+              secretName: flux-trust-store
+    target:
+      kind: Deployment
+      name: "(kustomize|image-automation|source|image-reflector|helm|notification)-controller"
+```
+
+#### Object-level Trusted Certificates Expansion
+
+A new field is to be introduced into the existing kinds `Bucket`, `GitRepository`,
+`HelmRepository`, `OCIRepository`, `ImageUpdateAutomation`, `Provider` and
+`ImageRepository`, to allow users to expand on the controller-level known hosts for
+SSH operations:
+
+```yaml
+spec:
+  trustStore:
+    tls:
+        secretRef:
+        configMapRef:
+```
+
+The trust store can be expanded by either setting `spec.trustStore.tls.secretRef`
+or `spec.trustStore.tls.configMapRef`, not both. Either option should contain the
+data under a `caFile` key.
+
+CA bundles configured this way will be aggregated with the ones defined at both
+system and controller levels.
+
+### Enabling Object-Level Trust Store
+
+Object-level trust store expansion is disabled by default. To enable it start
+the controller with:
```

Comment:
I would love to see this expanded a little to provide reasoning for this default setting. With the current Flux version 0.37 the trust store can always be expanded for objects such as HelmRepositories or GitRepositories by defining a `caFile` field in the referenced Secret. Given many admins will unlikely deviate from the defaults, users would no longer be able to do that. Please correct me if I'm wrong here.

Reply:
This is certainly a point up for debate. The security invariant here was "Flux should only connect to trusted servers" and the criteria used was that most users would expect Flux to be secure by default.

> Given many admins will unlikely deviate from the defaults

This is exactly the point. If something works, users won't look up the documentation and try understand the impact of security sensitive settings. Hence why ideally we should always prioritise security by design and by default.

The idea is that, by pre-populating the controller level SSH trust store with the top SaaS Git providers, a considerable amount of users would not have to worry about this. It will simply securely work - just like for TLS.
Then enterprise customers should configure this once, at controller level and be done with it, for both single and multi-tenancy. The only use case I see for Object-level trust stores to be used **in production**, would be in a Namespace as a Service multi-tenancy model, which would further weaken an already fragile security posture.